include/envoy/network/socket.h

```diff
@@ -43,22 +43,39 @@ struct SocketOptionName {
   Network::SocketOptionName(level, option, #level "/" #option)
 
 /**
- * Base class for Sockets
+ * Interfaces for providing a socket's various addresses. This is split into a getters interface
+ * and a getters + setters interface. This is so that only the getters portion can be overridden
+ * in certain cases.
  */
-class Socket {
+class SocketAddressProviderGetters {
 public:
-  virtual ~Socket() = default;
+  virtual ~SocketAddressProviderGetters() = default;
 
   /**
-   * Type of sockets supported. See man 2 socket for more details
+   * @return the local address of the socket.
    */
-  enum class Type { Stream, Datagram };
+  virtual const Address::InstanceConstSharedPtr& localAddress() const PURE;
 
   /**
-   * @return the local address of the socket.
+   * @return true if the local address has been restored to a value that is different from the
+   *         address the socket was initially accepted at.
    */
-  virtual const Address::InstanceConstSharedPtr& localAddress() const PURE;
+  virtual bool localAddressRestored() const PURE;
+
+  /**
+   * @return the remote address of the socket.
+   */
+  virtual const Address::InstanceConstSharedPtr& remoteAddress() const PURE;
+
+  /**
+   * @return the direct remote address of the socket. This is the address of the directly
+   *         connected peer, and cannot be modified by listener filters.
+   */
+  virtual const Address::InstanceConstSharedPtr& directRemoteAddress() const PURE;
+};
 
+class SocketAddressProvider : public SocketAddressProviderGetters {
```

Comment:
I mildly prefer names `SocketAddressProvider` and `SocketAddressSetter`, as `Provider` seems like a synonym for `Getter`.

Reply:
SGTM